controllers/smartlimiter_controller.go

```diff
@@ -120,32 +129,130 @@ func (r *SmartLimiterReconciler) SetupWithManager(mgr ctrl.Manager) error {
+
+	// TODO 0125 if enable query metric from prometheus ...
+	pc, err := newProducerConfig(env)
 	if err != nil {
-		log.Errorf("failed to create slime-metric,%+v", err)
+		log.Errorf("%v", err)
 		return nil
 	}
-	src.AppendSource(ms)
-	f := ms.SourceClusterHandler()
-	if cfg.Multicluster {
-		mc := multicluster.New(env, []func(*kubernetes.Clientset){f}, nil)
-		go mc.Run()
-	}
 	r := &SmartLimiterReconciler{
 		Client:               mgr.GetClient(),
 		scheme:               mgr.GetScheme(),
 		metricInfo:           cmap.New(),
-		eventChan:            eventChan,
-		source:               src,
+		interest: 			  cmap.New(),
 		env:                  env,
 		lastUpdatePolicyLock: &sync.RWMutex{},
-		//globalRateLimitInfo: cmap.New(),
+		watcherMetricChan:    pc.WatcherProducerConfig.MetricChan,
+		tickerMetricChan:     pc.TickerProducerConfig.MetricChan,
 	}
-	r.source.Start(env.Stop)
-	r.WatchSource(env.Stop)
+
+	// register handler in framework, and trigger by event or ticker
+	pc.WatcherProducerConfig.NeedUpdateMetricHandler = r.handleWatcherEvent
+	pc.TickerProducerConfig.NeedUpdateMetricHandler = r.handleTickerEvent
+	// start producer
+	metric.NewProducer(pc)
+	log.Infof("producers starts")
+
+
+	// TODO 0124 如果开关没有打开那么
+	if env.Config.Metric != nil {
+		go r.WatchMetric()
+	}
+
 	return r
 }
+
+//func NewReconciler(cfg *v1alpha1.Limiter, mgr ctrl.Manager, env *bootstrap.Environment) *SmartLimiterReconciler {
+//	log := log.WithField("controllers", "SmartLimiter")
+//	eventChan := make(chan event_source.Event)
+//	src := &aggregate.Source{}
+//	ms, err := k8s.NewMetricSource(eventChan, env)
+//	if err != nil {
+//		log.Errorf("failed to create slime-metric,%+v", err)
+//		return nil
+//	}
+//	src.AppendSource(ms)
+//	f := ms.SourceClusterHandler()
+//	if cfg.Multicluster {
+//		mc := multicluster.New(env, []func(*kubernetes.Clientset){f}, nil)
+//		go mc.Run()
+//	}
+//	r := &SmartLimiterReconciler{
+//		Client:               mgr.GetClient(),
+//		scheme:               mgr.GetScheme(),
+//		metricInfo:           cmap.New(),
+//		eventChan:            eventChan,
+//		source:               src,
+//		env:                  env,
+//		lastUpdatePolicyLock: &sync.RWMutex{},
+//		//globalRateLimitInfo: cmap.New(),
+//	}
+//	r.source.Start(env.Stop)
+//	r.WatchSource(env.Stop)
+//	return r
+//}
+
+
+// producerConfig 定义了
+
+func newProducerConfig(env bootstrap.Environment) (*metric.ProducerConfig, error) {
+
+	// init metric source
+	var enablePrometheusSource bool
+	var prometheusSourceConfig metric.PrometheusSourceConfig
+	var err error
+
+	// TODO 0124
+	switch env.Config.Global.Misc[model.MetricSourceType] {
+	case model.MetricSourceTypePrometheus:
+		enablePrometheusSource = true
+		prometheusSourceConfig, err = newPrometheusSourceConfig(env)
+		if err != nil {
+			return nil, err
+		}
+	default:
+		return nil, stderrors.New("wrong metric_source_type")
+	}
+
+	// TODO do not need to query prometheus 0124
```

Comment:
it seems like these `TODO` have relations with each other, maybe it's a hint for future tasks, make it more clearly for other people?